File: modules/shared/config/claude/hooks/git-commit-validator.py
#!/usr/bin/env python3
"""
Git commit validator hook for Claude Code
Prevents use of --no-verify flag by intercepting git commit commands
"""
import json
import sys
import re

try:
    input_data = json.load(sys.stdin)
except json.JSONDecodeError as e:
    print(f"Error: Invalid JSON input: {e}", file=sys.stderr)
    sys.exit(1)

tool_name = input_data.get("tool_name", "")
tool_input = input_data.get("tool_input", {})
command = tool_input.get("command", "")

# Only process Bash tool with git commit commands
if tool_name != "Bash":
    sys.exit(0)

# Check if this is a git commit command
if not re.search(r"\bgit\s+commit\b", command):
    sys.exit(0)

# Check for --no-verify flag
if re.search(r"--no-verify", command):
    print("⚠️  --no-verify 플래그 사용이 감지되었습니다.", file=sys.stderr)
    print("pre-commit hooks는 코드 품질을 보장하기 위해 중요합니다.", file=sys.stderr)
    print("대신 다음을 시도해보세요:", file=sys.stderr)
    print("1. pre-commit 오류를 수정하여 정상적으로 커밋", file=sys.stderr)
    print("2. 특정 hook만 건너뛰기: SKIP=ruff git commit -m '...'", file=sys.stderr)
    print("3. 정말 필요한 경우 수동으로 터미널에서 실행", file=sys.stderr)
    
    # Exit code 2 blocks the command and shows stderr to Claude
    sys.exit(2)

# Allow the command to proceed
sys.exit(0)

File: modules/shared/config/claude/hooks-go/internal/hooks/pretooluse/git_commit_validator.go
// Package pretooluse implements PreToolUse hooks that execute before tool calls
package pretooluse

import (
	"context"
	"regexp"

	"github.com/baleen/dotfiles/hooks-go/internal/hook"
)

// GitCommitValidator prevents use of --no-verify flag in git commit commands
type GitCommitValidator struct{}

// NewGitCommitValidator creates a new git commit validator hook
func NewGitCommitValidator() *GitCommitValidator {
	return &GitCommitValidator{}
}

// Name returns the hook identifier
func (h *GitCommitValidator) Name() string {
	return "git-commit-validator"
}

// Validate checks if git commit command contains --no-verify flag
func (h *GitCommitValidator) Validate(ctx context.Context, input *hook.Input) (*hook.Response, error) {
	resp := hook.NewResponse()

	// Only process Bash tool
	if input.ToolName != "Bash" {
		return resp, nil
	}

	// Get command from tool input
	commandInterface, ok := input.ToolInput["command"]
	if !ok {
		return resp, nil
	}

	command, ok := commandInterface.(string)
	if !ok {
		return resp, nil
	}

	// Check if this is a git commit command
	gitCommitPattern := regexp.MustCompile(`^\s*git\s+commit\b`)
	if !gitCommitPattern.MatchString(command) {
		return resp, nil
	}

	// Remove quoted content to avoid false positives
	cleanCommand := removeQuotedContent(command)

	// Check for --no-verify flag
	noVerifyPattern := regexp.MustCompile(`--no-verify\b`)
	if noVerifyPattern.MatchString(cleanCommand) {
		resp.Block("⚠️  --no-verify flag detected.\n" +
			"Pre-commit hooks are important for code quality.\n" +
			"Try these alternatives instead:\n" +
			"1. Fix pre-commit errors and commit normally\n" +
			"2. Skip specific hooks: SKIP=ruff git commit -m '...'\n" +
			"3. Run manually in terminal if absolutely necessary")
		return resp, nil
	}

	return resp, nil
}

// removeQuotedContent removes content within quotes to avoid false positives
func removeQuotedContent(s string) string {
	// Remove double-quoted strings
	doubleQuotePattern := regexp.MustCompile(`"[^"]*"`)
	s = doubleQuotePattern.ReplaceAllString(s, "")

	// Remove single-quoted strings
	singleQuotePattern := regexp.MustCompile(`'[^']*'`)
	s = singleQuotePattern.ReplaceAllString(s, "")

	return s
}
